src/hotspot/share/memory/metaspace.hpp

```diff
@@ -113,8 +117,30 @@ class Metaspace : public AllStatic {
   static MetaWord* allocate(ClassLoaderData* loader_data, size_t word_size,
                             MetaspaceObj::Type type);
 
-  static bool contains(const void* ptr);
-  static bool contains_non_shared(const void* ptr);
+  static bool contains(const void* ptr) {
+    return is_in_shared_metaspace(ptr) || // in cds
+           is_in_class_space(ptr) ||      // in class space
+           is_in_nonclass_metaspace(ptr); // in one of the non-class regions?
+  }
+
+  // kept for now for backward compat reasons, but lets test if callers really need this
+  static bool contains_non_shared(const void* ptr) {
+    return is_in_class_space(ptr) ||      // in class space
+           is_in_nonclass_metaspace(ptr); // in one of the non-class regions?
+  }
+
+  // Returns true if pointer points into one of the metaspace regions, or
+  // into the class space.
+  static bool is_in_shared_metaspace(const void* ptr);
+
+  // Returns true if pointer points into one of the non-class-space metaspace regions.
+  static bool is_in_nonclass_metaspace(const void* ptr);
+
+  // Returns true if ptr points into class space, false if it doesn't or if
+  // there is no class space.
```

Comment:
Add a comment that the class space is one contiguous allocated region, which is why this pointer compare is correct, unlike the non-class metaspace.